### DISCUSSION TITLE
[Proposal]: Nullable analysis of LINQ queries

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#8998] [General]
# Nullable analysis of LINQ queries

Champion issue: https://github.com/dotnet/csharplang/issues/8999

## Summary
[summary]: #summary

Nullable analysis of LINQ queries (LDM expressed interested to handle `Where`, needs design proposal)

## Motivation
[motivation]: #motivation

Range variables in LINQ queries are currently treated as oblivious to minimize annoyances.
Most of the annoyance results from our inability to recognize method calls that filter out `null` elements, such as `.Where(i => i is not null)`.

```
using System.Linq;
using System.Collections.Generic;

#nullable enable
public class C 
{
    public void M(IEnumerable<string?> list) 
    {
        _ = from item in list
            select item.ToString(); // should warn

        _ = list
            .Where(i => i is not null)
            .Select(i => i.ToString()); // should not warn
    }
}
```

## Detailed design
[design]: #detailed-design

TBD

## Unresolved questions
[unresolved]: #unresolved-questions

- [ ] How to recognize `Where` methods?

## Design meetings

Split issue from https://github.com/dotnet/csharplang/issues/3868
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-09-28.md#nullability-improvements
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-09-06.md#nullable-analysis-of-linq-queries

**Comment by gafter** [comment]
Would these warnings only be suppressed for the "standard" Linq implementations that have the expected semantics, or for all query expressions (even those that do not have the expected semantics)?

**Comment by JTeeuwissen** [comment]
Note that `OfType` currently can be used to filter out nulls whilst getting the correct return type. But perhaps there are more cases where this would not apply.

**Comment by tobias-tengler** [comment]
What's the status of this proposal?

**Comment by KillyMXI** [comment]
Last meeting note link is wrong. Should be this, it seems: https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-09-04.md#nullable-analysis-of-linq-queries

**Comment by JudahGabriel** [comment]
Slightly related: TypeScript [added support for this](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#inferred-type-predicates) last year:

```typescript
var birds = nations
   .map(n => n.getNationalBird()) // getNationalBird returns Bird or null
   .filter(n => n != null);

birds.forEach(bird => bird.sing()); // Typescript understands bird will be non-null here
```

It's found great use within the TS community. I suspect the same would be true of C#.
